src/engine/graphics/shaders.cpp

```diff
@@ -0,0 +1,210 @@
+	} catch ( const std::exception& error ) {
+		LogWarn( "Failed to register shader program, %s!\n", error.what());
+	}
+}
+
+static void Shaders_ClearPrograms() {
+	for ( const auto& program : programs ) {
+		program.second->Disable();
+		delete program.second;
+	}
+
+	programs.clear();
+}
+
+static void Shaders_CachePrograms() {
+	Shaders_ClearPrograms();
+
+	char scanPath[PL_SYSTEM_MAX_PATH];
+	snprintf( scanPath, sizeof( scanPath ), "%s/shaders", u_get_base_path());
+	plScanDirectory( scanPath, "program", Shaders_CacheShaderProgram, false );
+
+	// Recursive searches through dependencies... This needs revisiting!
+	const char* modPath = u_get_mod_path();
+	if ( modPath[ 0 ] != '\0' ) {
+		snprintf( scanPath, sizeof( scanPath ), "%s/shaders", u_get_full_path());
+		plScanDirectory( scanPath, "program", Shaders_CacheShaderProgram, false );
+	}
+
+	Shaders_ValidateDefault();
+}
+
+static void Cmd_RebuildShaderPrograms( unsigned int argc, char* argv[] ) {
+	Shaders_CachePrograms();
+}
+
+static void Cmd_RebuildShaderProgram( unsigned int argc, char* argv[] ) {
+	u_unused( argc );
+	u_unused( argv );
+
+	try {
+		for ( const auto& program : programs ) {
+			program.second->Rebuild();
+		}
+	} catch ( ... ) {
+		Error( "Failed to rebuild shader programs, check logs for details!\n" );
+	}
+
+	Shaders_SetProgramByName( lastProgramName );
+}
+
+void Shaders_Initialize() {
+	plRegisterConsoleCommand( "rebuildShaderPrograms", Cmd_RebuildShaderPrograms, "Rebuild all shader programs" );
+	plRegisterConsoleCommand( "rebuildShaderProgram", Cmd_RebuildShaderProgram, "Rebuild specified shader program" );
+
+	Shaders_CachePrograms();
+}
+
+void Shaders_Shutdown() {
+	Shaders_ClearPrograms();
+}
+
+hwShaderProgram* Shaders_GetProgram( const std::string& name ) {
+	const auto& i = programs.find( name );
+	if ( i == programs.end()) {
+		LogWarn( "Failed to find shader program, \"%s\"!\n", name.c_str());
+		return nullptr;
+	}
+
+	return i->second;
+}
+
+void Shaders_SetProgram( hwShaderProgram* shaderProgram ) {
+	shaderProgram->Enable();
+}
+
+void Shaders_SetProgramByName( const std::string& name ) {
+	hwShaderProgram* shaderProgram = Shaders_GetProgram( name );
+	if ( shaderProgram == nullptr ) {
+		shaderProgram = fallbackShaderProgram;
+	}
+
+	Shaders_SetProgram( shaderProgram );
+}
+
+hwShaderProgram::hwShaderProgram( const std::string& vertPath, const std::string& fragPath ) {
+	shaderProgram = plCreateShaderProgram();
+	if ( shaderProgram == nullptr ) {
+		throw std::runtime_error( plGetError());
+	}
+
+	try {
+		RegisterShaderStage( vertPath.c_str(), PL_SHADER_TYPE_VERTEX );
+		RegisterShaderStage( fragPath.c_str(), PL_SHADER_TYPE_FRAGMENT );
+	} catch ( ... ) {
+		plDestroyShaderProgram( shaderProgram, true );
+		throw;
+	}
+
+	plLinkShaderProgram( shaderProgram );
+
+	this->vertPath = vertPath;
+	this->fragPath = fragPath;
+}
+
+hwShaderProgram::~hwShaderProgram() {
+	plDestroyShaderProgram( shaderProgram, true );
+}
+
+void hwShaderProgram::Rebuild() {
```

Comment:
Suggest changing this to set up new shader program before deleting and replacing shaderProgram.